src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+ *  <ul>
+ *      <li>When this occurs the warning displays each step in the stack trace to help in comprehension.
+ *  </ul>
+ *  <li>We assume that native methods do not leak.
+ *  <li>We don't try to follow {@code super()} invocations; that's for the superclass analysis to handle.
+ *  </ul>
+ */
+class ThisEscapeAnalyzer extends TreeScanner {
+
+    private final Names names;
+    private final Types types;
+    private final Log log;
+    private       Lint lint;
+
+// These fields are scoped to the entire COMPILATION UNIT
+
+    /** Maps symbols of all methods to their corresponding declarations.
+     */
+    private final Map<Symbol, MethodInfo> methodMap = new LinkedHashMap<>();
+
+    /** Contains symbols of fields and methods that have warnings suppressed.
+     */
+    private final Set<Symbol> suppressed = new HashSet<>();
+
+    /** The declaring class of the constructor we're currently analyzing.
+     *  This is the 'this' type we're trying to detect leaks of.
+     */
+    private JCClassDecl targetClass;
+
+    /** Snapshots of {@link #callStack} where possible 'this' escapes occur.
+     */
+    private ArrayList<DiagnosticPosition[]> warningList = new ArrayList<>();
+
+// These fields are scoped to the CONSTRUCTOR BEING ANALYZED
+
+    /** The declaring class of the "invoked" method we're currently analyzing.
+     *  This is either the analyzed constructor or some method it invokes.
+     */
+    private JCClassDecl methodClass;
+
+    /** The current "call stack" during our analysis. The first entry is some method
+     *  invoked from the target constructor; if empty, we're still in the constructor.
+     */
+    private final ArrayDeque<DiagnosticPosition> callStack = new ArrayDeque<>();
+
+    /** Used to terminate recursion in {@link #invokeInvokable invokeInvokable()}.
+     */
+    private final Set<Pair<JCTree, RefSet<Ref>>> invocations = new HashSet<>();
+
+    /** Snapshot of {@link #callStack} where a possible 'this' escape occurs.
+     *  If non-null, a 'this' escape warning has been found in the current
+     *  constructor statement, initialization block statement, or field initializer.
+     */
+    private DiagnosticPosition[] pendingWarning;
+
+// These fields are scoped to the CONSTRUCTOR OR INVOKED METHOD BEING ANALYZED
+
+    /** Current lexical scope depth in the constructor or method we're currently analyzing.
+     *  Depth zero is the outermost scope. Depth -1 means we're not analyzing.
+     */
+    private int depth = -1;
+
+    /** Possible 'this' references in the constructor or method we're currently analyzing.
+     *  Null value means we're not analyzing.
+     */
+    private RefSet<Ref> refs;
+
+// Constructor
+
+    ThisEscapeAnalyzer(Names names, Types types, Log log, Lint lint) {
+        this.names = names;
+        this.types = types;
+        this.log = log;
+        this.lint = lint;
+    }
+
+//
+// Main method
+//
+
+    public void analyzeTree(Env<AttrContext> env) {
```

Comment:
nit: this method could be removed in favor of the overloaded version below

Reply:
Thanks - will fix.